src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecutionTest.java

```diff
@@ -527,4 +527,26 @@ public boolean isAllowed(String groovyResourceUrl) {
             return groovyResourceUrl.endsWith("/trusted/foo.groovy");
         }
     }
+
+    @Issue("JENKINS-45327")
+    @Test public void envActionImplPickle() throws Throwable {
+        sessions.then(r -> {
+            WorkflowJob p = r.createProject(WorkflowJob.class, "p");
+            p.setDefinition(new CpsFlowDefinition(
+                    "def e = env\n" +
+                    "semaphore('wait')\n" + // An instance of EnvActionImpl is part of the program's state at this point.
+                    "e.foo = 'bar'\n" + // Without EnvActionImplPickle, this throws an NPE in EnvActionImpl.setProperty because owner is null.
+                    "echo('EnvActionImpl instance: ' + System.identityHashCode(e))\n", false));
+            WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+            SemaphoreStep.waitForStart("wait/1", b);
+        });
+        sessions.then(r -> {
+            WorkflowJob p = r.jenkins.getItemByFullName("p", WorkflowJob.class);
+            WorkflowRun b = p.getLastBuild();
+            SemaphoreStep.success("wait/1", null);
+            r.assertBuildStatus(Result.SUCCESS, r.waitForCompletion(b));
+            // Check that the EnvActionImpl attached to the run is the same object as the one in the program.
```

Comment:
Rather than this low-level check, why not simply assert that the final env includes `foo=bar`, which is the actual goal?

Reply:
Yeah good point. At first I just wanted to show that the instance was the same, but checking the value of `foo` on the action attached to the run does that indirectly anyway and tests the user-facing behavior which seems better, so I updated the test in 06761ef2c7b6aac6da9659476ca93feed3d0238c.